### PR TITLE
Don't print warning messages for DisplayServer functions in headless mode

### PR DESCRIPTION
**Pull request by Calinou**
DisplayServer methods that only have a cosmetic effect print a warning (instead of an error). These warnings can be silenced in headless mode, as it's assumed that the cosmetic effect is undesired in headless mode anyway (e.g. changing mouse cursor shape).

This prevents spurious warnings from appearing when running the editor in headless mode, e.g. on CI to export a project.

Methods that print an error will continue to do so, as their returned value may not match expectations and have cascading consequences on the project (e.g. clipboard data not being retrievable).

- See [Supressing mouse warnings from running server in –headless mode - Godot Forum](https://forum.godotengine.org/t/supressing-mouse-warnings-from-running-server-in-headless-mode/54116).

**Testing project:** [test_headless_methods.zip](https://github.com/godotengine/godot/files/14951654/test_headless_methods.zip)
Run with and without `--headless` from a terminal and compare output.

### Before

```
$ godot --headless 
Godot Engine v4.3.dev.custom_build.b091162a9 (2024-04-10 15:49:59 UTC) - https://godotengine.org

WARNING: Custom cursor shape not supported by this display server.
     at: cursor_set_custom_image (servers/display_server.cpp:636)
WARNING: Clipboard is not supported by this display server.
     at: clipboard_set (servers/display_server.cpp:506)
ERROR: Clipboard is not supported by this display server.
   at: clipboard_get (servers/display_server.cpp:510)

ERROR: Virtual keyboard not supported by this display server.
   at: virtual_keyboard_get_height (servers/display_server.cpp:624)
0
WARNING: Native dialogs not supported by this display server.
     at: dialog_input_text (servers/display_server.cpp:652)
```

### After

```
$ godot --headless
Godot Engine v4.3.dev.custom_build.67b915f63 (2024-04-11 21:14:41 UTC) - https://godotengine.org

ERROR: Clipboard is not supported by this display server.
   at: clipboard_get (servers/display_server.cpp:517)

ERROR: Virtual keyboard not supported by this display server.
   at: virtual_keyboard_get_height (servers/display_server.cpp:631)
0
```

### Non-headless output for comparison

*This is left unchanged in this PR.*

```
$ godot           
Godot Engine v4.3.dev.custom_build.67b915f63 (2024-04-11 21:14:41 UTC) - https://godotengine.org
Vulkan 1.3.277 - Forward+ - Using Device #0: NVIDIA - NVIDIA GeForce RTX 4090

test
ERROR: Virtual keyboard not supported by this display server.
   at: virtual_keyboard_get_height (servers/display_server.cpp:631)
0
WARNING: Native dialogs not supported by this display server.
     at: dialog_input_text (servers/display_server.cpp:663)
```

### CODE REVIEW

**Comment by m4gr3d** [comment]
> Sorry, wouldn't it be simpler to just implement ~a stub method~ stub methods in the headless class? Since it's an exception, I think that a local solution would be better.
> 
> Edit: oops I said "stub method" but I meant like, stubs in general.

I agree; this breaks encapsulation principle if the parent / base class has special logic for one of its derived class.

**Review by m4gr3d — CHANGES_REQUESTED** [review]
Added feedback in comments.

**Comment by Calinou** [comment]
> Sorry, wouldn't it be simpler to just implement ~a stub method~ stub methods in the headless class? Since it's an exception, I think that a local solution would be better.
> 
> Edit: oops I said "stub method" but I meant like, stubs in general.

Done. I've tested the MRP again and I get the same output as before.

File: servers/display_server_headless.h [hunk 1]
```diff
@@ -134,6 +134,57 @@ class DisplayServerHeadless : public DisplayServer {
...
+	virtual IndicatorID create_status_indicator(const Ref<Image> &p_icon, const String &p_tooltip, const Callable &p_callback) override { return INVALID_INDICATOR_ID; }
+	virtual void status_indicator_set_icon(IndicatorID p_id, const Ref<Image> &p_icon) override {}
+	virtual void status_indicator_set_tooltip(IndicatorID p_id, const String &p_tooltip) override {}
+	virtual void status_indicator_set_callback(IndicatorID p_id, const Callable &p_callback) override {}
+	virtual void delete_status_indicator(IndicatorID p_id) override {}
+
+	virtual int64_t window_get_native_handle(HandleType p_handle_type, WindowID p_window = MAIN_WINDOW_ID) const override { return 0; }
+
+	virtual void window_set_vsync_mode(VSyncMode p_vsync_mode, WindowID p_window = MAIN_WINDOW_ID) override {}
+	virtual VSyncMode window_get_vsync_mode(WindowID p_window) const override { return VSyncMode::VSYNC_ENABLED; }
```
Comment (m4gr3d) [inline servers/display_server_headless.h:186]: Should this be `VSYNC_DISABLED`?

**Review by m4gr3d — APPROVED** [review]

File: servers/display_server_headless.h (hunk 1 above)
Comment (Calinou) [reply servers/display_server_headless.h:186]: The fallback value is `VSYNC_ENABLED`: https://github.com/godotengine/godot/blob/029aadef563fb69cf49aa9795b62f27171f8c3f4/servers/display_server.cpp#L748

File: servers/display_server_headless.h [hunk 2]
```diff
@@ -130,10 +135,48 @@ class DisplayServerHeadless : public DisplayServer {
...
 
+	void set_native_icon(const String &p_filename) override {}
 	void set_icon(const Ref<Image> &p_icon) override {}
 
+	void help_set_search_callbacks(const Callable &p_search_callback = Callable(), const Callable &p_action_callback = Callable()) override {}
+
+	bool tts_is_speaking() const override { return false; }
+	bool tts_is_paused() const override { return false; }
+	TypedArray<Dictionary> tts_get_voices() const override { return TypedArray<Dictionary>(); }
+	void tts_speak(const String &p_text, const String &p_voice, int p_volume = 50, float p_pitch = 1.f, float p_rate = 1.f, int p_utterance_id = 0, bool p_interrupt = false) override {}
```
Comment (aaronfranke) [inline servers/display_server_headless.h:150]:
```suggestion
	void tts_speak(const String &p_text, const String &p_voice, int p_volume = 50, float p_pitch = 1.0f, float p_rate = 1.0f, int p_utterance_id = 0, bool p_interrupt = false) override {}
```

**Review by aaronfranke — APPROVED** [review]
This isn't my area and I haven't tested it but the change is very straightforward and looks good to me.

**Comment by akien-mga** [comment]
I'm not against adding stubs, but at least some of those may need to be guarded with `DisplayServer::get_singleton()->has_feature()` checks before being called (which means that actually showing a warning when called should be kept, as it's information for end users when they themselves wrongly rely on the API).

So it might be worth reviewing places in the engine that call some of those APIs without guarding with the proper feature check.

For example:
```
WARNING: Custom cursor shape not supported by this display server.
     at: cursor_set_custom_image (servers/display_server.cpp:636)
```
Shouldn't happen anymore already after e6d0bf32e9cfca1118d95a371d869c8ec0d43d2b.

`DisplayServerHeadless::has_feature()` is always `false` so in theory it should be sufficient.

**Comment by akien-mga** [comment]
If someone wants to look into which checks may actually be missing in the core engine before merging this (see my above comment), that could be good.

But otherwise I'm also fine merging this as is if there's no objection from @bruvzg on making these no-ops (so end users won't know why they don't work on platforms that don't support them, and we need to make sure the docs flag those properly).

**Review by bruvzg — COMMENTED** [review]
Adding stubs to the headless DS should be fine.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**